module/zfs/blake3_zfs.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://opensource.org/licenses/CDDL-1.0.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright 2021 Tino Reichardt <milky-zfs@mcmilk.de>
+ */
+
+#include <sys/zfs_context.h>
+#include <sys/zio.h>
+#include <sys/zio_checksum.h>
+#include <sys/blake3.h>
+#include <sys/abd.h>
+
+static int
+blake3_incremental(void *buf, size_t size, void *arg)
+{
+	BLAKE3_CTX *ctx = arg;
+	Blake3_Update(ctx, buf, size);
+	return (0);
+}
+
+/*
+ * Computes a native 256-bit BLAKE3 MAC checksum. Please note that this
+ * function requires the presence of a ctx_template that should be allocated
+ * using abd_checksum_blake3_tmpl_init.
+ */
+void
+abd_checksum_blake3_native(abd_t *abd, uint64_t size, const void *ctx_template,
+    zio_cksum_t *zcp)
+{
+	/*
+	 * CC [M]  module/zfs/blake3_zfs.o
+	 * module/zfs/blake3_zfs.c: In function ‘abd_checksum_blake3_native’:
+	 * module/zfs/blake3_zfs.c:56:1: warning: the frame size of 1920 bytes
+	 *     is larger than 1024 bytes [-Wframe-larger-than=]
+	 */
+	BLAKE3_CTX ctx;
```

Comment:
Yeah, that large stack variable may be a problem, you might need to switch to kmem_cache_alloc or similar.

Reply:
Yes, I got the warning also - but not an all machines... thats why I forgot about it again ;)
I will try the switch to mem_cache_alloc() and will take a look, if it performs ...

Reply:
- feature flag ist done
- How will that impact the send stream format? -> the Hash will be used like edonr or skein (mainly dedup)
- kmem_cache_alloc() works nicely

Reply:
Hey @allanjude, I have no idea why the FreeBSD tests are failing in the moment. They tell that Blake3_Init(), Blake3_Update() and Final() can not be found. But the ipc.ko module does export them.

Is there a need to create module/icp/io/blake3_mod.c or sth. ?
In general - should I provide this module - is it needed ?

Thanks for your help.

Reply:
FreeBSD doesn't build icp by default. You'll need to add the .c files you want to the list in Makefile.BSD to compile them into the main kernel module (see what @rincebrain did recently for edon-r)